--- client.py	2017-07-26 10:11:26.000000000 -0700
+++ client.py.new	2017-07-26 10:11:21.000000000 -0700
@@ -10,7 +10,7 @@
 # distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
 # ANY KIND, either express or implied. See the License for the specific
 # language governing permissions and limitations under the License.
-import logging
+import logging,os,boto3,time,uuid
 from datetime import datetime, timezone
 import functools
 
@@ -307,6 +307,61 @@
             if args:
                 raise TypeError(
                     "%s() only accepts keyword arguments." % py_operation_name)
+
+            # SpotWrap additions to capture AWS API calls
+            reqID = 'unknown'
+            if 'spotReqID' in os.environ: #this will be set to 'unknown' if outside of lambda
+                reqID = os.environ['spotReqID']
+            #skip if its a write to the dynamodb spot table (infinite loop)
+            if not(reqID == 'unknown' or (operation_name == 'PutItem' and kwargs['TableName'] == 'spotFns')): 
+                myArn = 'unknown'
+                if 'myArn' in os.environ:
+                    myArn = os.environ['myArn']
+                dynamodb = boto3.resource('dynamodb', region_name='us-west-2')
+                table = dynamodb.Table('spotFns')
+                msg = 'SW'
+                eventOp = 'unknown'
+                if operation_name == 'PutObject' or operation_name == 'GetObject': #s3
+                    msg +=  ':{}:{}'.format(kwargs['Bucket'],kwargs['Key'])
+                    #also kwargs['Body'] for Put
+                    if operation_name == 'GetObject':
+                        eventOp = 'os_read'
+                    if operation_name == 'PutObject':
+                        eventOp = 'os_write'
+                elif operation_name == 'PutItem' or operation_name == 'GetItem': #dynamodb
+                    msg +=  ':{}'.format(kwargs['TableName'])
+                    if operation_name == 'GetItem':
+                        msg +=  ':{}'.format(str(kwargs['Key']))
+                        eventOp = 'db_read'
+                    else:
+                        msg +=  ':{}'.format(str(kwargs['Item']))
+                        eventOp = 'db_write'
+                elif operation_name == 'Invoke':
+                    msg +=  ':{}'.format(kwargs['FunctionName'])
+                    eventOp = 'invoke'
+		    #also: kwargs['Payload']
+                else:
+                    msg += 'SpotWrapError:unknown_op'
+                reqID += ':{}:{}'.format(operation_name,str(uuid.uuid4())[:8])
+                ts = time.time() * 1000 #msecs
+                tsint = int(round(ts))
+                table.put_item( Item={
+                    'ts': tsint,
+                    'requestID': reqID,
+                    'thisFnARN': myArn,
+                    'caller': 'unknown',
+                    'eventSource': 'unknown',
+                    'eventOp': eventOp,
+                    'region': 'unknown',
+                    'accountID': 'unknown',
+                    'sourceIP': 'unknown',
+                    'message': msg,
+                    'duration': 0,
+                    'error': 'unknown',
+                    }
+               )
+
+
             # The "self" in this scope is referring to the BaseClient.
             return self._make_api_call(operation_name, kwargs)
 
